### PR TITLE
Don't show stack outputs when update fails

### PR DESCRIPTION
**Pull request by swgillespie**
It is basically guaranteed that stack outputs are going to be
meaningless if a plan fails, so this commmit doesn't display them if an
error has occured.

Fixes pulumi/pulumi#1913, in that it won't show outputs in Luke's example.

### CODE REVIEW

File: pkg/backend/display/progress.go [hunk 1]
```diff
@@ -709,8 +716,8 @@ func (display *ProgressDisplay) processEndSteps() {
 		}
 	}
 
-	// If we get stack outputs, display them at the end.
-	if display.stackUrn != "" {
+	// If we get stack outputs and there weren't any errors, display them at the end.
+	if display.stackUrn != "" && !sawError {
```
Comment (CyrusNajmabadi) [inline pkg/backend/display/progress.go]: it's worthwhile in cases like this to explain the 'why' not the what.  i.e. why would we not want to show the error.

Comment (swgillespie) [reply pkg/backend/display/progress.go]: will do.

**Comment by swgillespie** [comment]
thanks!

**Comment by joeduffy** [comment]
If a program gets far enough to register a new one, or update one, isn't it worth displaying since the next update will then see if as no diff? Or am I misunderstand in?

**Comment by swgillespie** [comment]
> If a program gets far enough to register a new one, or update one, isn't it worth displaying since the next update will then see if as no diff? Or am I misunderstanding?

It's unlikely that this would happen in a way that is particularly interesting. Outputs are registered even if a program terminates due to an exception, but it's most likely that a user's program will fail with an exception during a preview, not the update (given that the vast majority of code runs outside of `apply`). It's guaranteed *not* to happen if any resource fails to register successfully, since the failing RPC will never return and eventually the CLI will just kill the language host.

**Comment by joeduffy** [comment]
Hmm, there are plenty of failures that occur during updates, but not previews.

If I say

```
export let a = foo();
export let b = bar();
```

and one of the two asynchronous calls, `foo` or `bar`, leads to an error during update (but not preview), then I'm pretty sure the situation I describe will occur, and I'm equally pretty sure the resulting impact to the output diff display is going to be confusing...

It seems better to simply not show deletions, because we cannot know that a computation that _would_ have produced the apparently-deleted output simply didn't get terminated midstream.

Again, maybe I'm wrong, but I do think this is going to be a problem.

**Comment by swgillespie** [comment]
> Hmm, there are plenty of failures that occur during updates, but not previews.

It's true that it's not hard to construct an example like this, but my point here is that stack outputs are registered only on normal and exceptional terminations of the language host - and not when `RegisterResource` fails. You'd have to construct a case where you either branch on `pulumi.isDryRun()` and then throw or do something that throws in an `apply`, which while possible I think are generally unlikely.

> It seems better to simply not show deletions, because we cannot know that a computation that would have produced the apparently-deleted output simply didn't get terminated midstream.

I agree with the latter part of this, and I think it's also a good argument as to why we shouldn't be showing stack outputs at all if there was an error - there's no way of knowing what sort of diff would have been produced had the program been allowed to continue. 

You're right of course in that is possible to write a program that fails, causes stack outputs to be updated, and doesn't see a diff in the CLI. However, the only way that I can think of that can cause this to happen is by throwing during an update - if there is even one `RegisterResource` that fails, the outputs will not be updated.

I'm happy to re-visit this if that seems confusing. I thought about this a bunch this morning and I wonder if we should instead *never register outputs at all* if a program is terminating with an exception. To me it breaks the model of "`pulumi:pulumi:Stack` is just a resource" to provide outputs if the resource itself failed to create successfully. Then again, we already challenge this model today by updating the stack while generating a `Same` step for `pulumi:pulumi:Stack`...

**Comment by swgillespie** [comment]
@joeduffy I'm also tracking down https://github.com/pulumi/pulumi/issues/1885, which is close enough to this issue that I will likely revert this fix and pursue a more complete one.

**Pull request merged**